### DISCUSSION TITLE
How to force unzoom for pane selection?

### DISCUSSION THREAD

**Discussion by twisteria** [wezterm/wezterm#2284] [Q&A] [answered]
Is there any actions defined to "zoom" or "unzoom" pane, not "toggle"(TogglePaneZoomState)?

I have created custom mode (key mappings) for switching panes with the option `one_shot=false` so that I could choose whether just simply moving the cursor or zooming into the pane when escaping from the mode. I put actual code below for your reference.
```
return {
	keys = {
		{key="f", mods="CTRL", action=act.ActivateKeyTable{
				name = "focus_mode",
				one_shot=false,
				replace_current=false,
		}}},
	key_tables = {
		focus_mode = {
			{key="h", action=act.ActivatePaneDirection("Left")},
			{key="j", action=act.ActivatePaneDirection("Down")},
			{key="k", action=act.ActivatePaneDirection("Up")},
			{key="l", action=act.ActivatePaneDirection("Right")},
			{key="LeftArrow",  action=act.ActivatePaneDirection("Left")},
			{key="DownArrow",  action=act.ActivatePaneDirection("Down")},
			{key="UpArrow",    action=act.ActivatePaneDirection("Up")},
			{key="RightArrow", action=act.ActivatePaneDirection("Right")},

			{key="f",     action="PopKeyTable"},
			{key="Enter", action="PopKeyTable"},
			{key="Space", action=act.Multiple{"PopKeyTable", act.TogglePaneZoomState}},
			{key="i",     action=act.Multiple{"PopKeyTable", act.SendKey{key="i"}}}, -- enabling insert mode for vim
		},
	}
}
```

My wish is to always start this "pane switching mode" with panes unzoomed -- if current state is unzoomed, start right away, if zoomed, unzoom and start the mode.

What I have come up so far is:
1: mapping something like `DisablePaneZoomState` for keys to cancel the mode, which I could not find such action.
2: code with lua's if statement, which I have little experience and have no clue
3: enable `unzoom_on_switch_pane` and always start pane switching mode after switching left and back to right as follows:
```
{key="f", mods="CTRL", action=act.Multiple{
	ActivatePaneDirection="Left",
	ActivatePaneDirection="Right",
	act.ActivateKeyTable{
		name = "focus_mode",
		one_shot=false,
		replace_current=false,
}}},
```
this, in fact, did nothing -- just started the mode.

Please help...

**Comment by wez** [comment, marked as answer]
There weren't but I just added some things in 112feb337461709e353f1ce4964d05873404be54:
[SetPaneZoomState](https://wezfurlong.org/wezterm/config/lua/keyassignment/SetPaneZoomState.html) key assignment and [MuxTab:set_zoomed()](https://wezfurlong.org/wezterm/config/lua/MuxTab/set_zoomed.html) for explicitly setting the zoom state of a pane.

It typically takes about an hour before changes are available as nightly builds for all platforms.  Linux builds are the fastest to build and are often available within about 20 minutes.  Windows and macOS builds take a bit longer.

Please take a few moments to try out the fix and let me know how that works out. You can find the nightly downloads for your system in the [wezterm installation docs](https://wezfurlong.org/wezterm/installation.html).

If you prefer to use packages provided by your distribution or package manager of choice and don't want to replace that with a nightly download, keep in mind that you can download portable packages (eg: a `.dmg` file on macOS, a `.zip` file on Windows and an `.AppImage` file on Linux) that can be run without permanently installing or replacing an existing package, and can then simply be deleted once you no longer need them. 

If you are eager and can [build from source](https://wezfurlong.org/wezterm/install/source.html) then you may be able to try this out more quickly.

**Reply by twisteria** [reply]
Thank you! it worked quite well!- also appreciate your detailed instruction on handling nightly!

Just for somebody's reference, below is the working config:
- Ctrl-f to switch to "pane selection" mode
- arrows/ h,j,k,l to navigate across panes
- arrows/ h,j,k,l with Shift key to change pane size
- f, enter, escape to finish the mode
- i to finish the mode and switch to insert mode in vim
- space to finish the mode with zoom to the selected pane
- d to delete a selected pane

```
return {
	keys = {
		{key="f", mods="CTRL", action=act.Multiple{
			act.SetPaneZoomState(false),
			act.ActivateKeyTable{
				name = "focus_mode",
				one_shot=false,
				replace_current=false,
			},
		}},
	},
	key_tables = {
		focus_mode = {
			{key="h", action=act.ActivatePaneDirection("Left")},
			{key="j", action=act.ActivatePaneDirection("Down")},
			{key="k", action=act.ActivatePaneDirection("Up")},
			{key="l", action=act.ActivatePaneDirection("Right")},
			{key="LeftArrow",  action=act.ActivatePaneDirection("Left")},
			{key="DownArrow",  action=act.ActivatePaneDirection("Down")},
			{key="UpArrow",    action=act.ActivatePaneDirection("Up")},
			{key="RightArrow", action=act.ActivatePaneDirection("Right")},

			{key="h", mods="SHIFT", action=act.AdjustPaneSize{"Left", 1}},
			{key="j", mods="SHIFT", action=act.AdjustPaneSize{"Down", 1}},
			{key="k", mods="SHIFT", action=act.AdjustPaneSize{"Up", 1}},
			{key="l", mods="SHIFT", action=act.AdjustPaneSize{"Right", 1}},
			{key="LeftArrow",  mods="SHIFT", action=act.AdjustPaneSize{"Left", 1}},
			{key="DownArrow",  mods="SHIFT", action=act.AdjustPaneSize{"Down", 1}},
			{key="UpArrow",    mods="SHIFT", action=act.AdjustPaneSize{"Up", 1}},
			{key="RightArrow", mods="SHIFT", action=act.AdjustPaneSize{"Right", 1}},

			{key="f",     action="PopKeyTable"},
			{key="Enter", action="PopKeyTable"},
			{key="Escape", action="PopKeyTable"},
			{key="d", action=act.CloseCurrentPane{confirm=false}},
			{key="Space", action=act.Multiple{"PopKeyTable", act.SetPaneZoomState(true)}},
			{key="i",     action=act.Multiple{"PopKeyTable", act.SendKey{key="i"}}},
		},
	},
}
```
